public_html/wp-content/themes/pattern-directory/src/components/report-pattern-modal/index.js

```diff
@@ -0,0 +1,155 @@
+/**
+ * WordPress dependencies
+ */
+import { useMemo, useReducer, useState } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+import { Button, Modal, RadioControl, Spinner, TextareaControl } from '@wordpress/components';
+import { useSelect } from '@wordpress/data';
+import apiFetch from '@wordpress/api-fetch';
+import { addQueryArgs } from '@wordpress/url';
+
+/**
+ * Internal dependencies
+ */
+import { store as patternStore } from '../../store';
+
+const reducer = ( state, action ) => {
+	switch ( action.status ) {
+		case 'submitting':
+			return { isSubmitting: true };
+		case 'submitted':
+			return { isSubmitted: true };
+		case 'error':
+			return { hasError: true, message: action.message };
+		default:
+			return {};
+	}
+};
+
+const ReportPatternModal = ( { postId, onClose } ) => {
+	const [ state, dispatch ] = useReducer( reducer, {} );
+	const [ selectedOption, setOption ] = useState( '' );
+	const [ details, setDetails ] = useState( '' );
+
+	const { isLoading, reasons } = useSelect( ( select ) => {
+		const { getPatternFlagReasons, isLoadingPatternFlagReasons } = select( patternStore );
+
+		return {
+			isLoading: isLoadingPatternFlagReasons(),
+			reasons: getPatternFlagReasons(),
+		};
+	} );
+
+	const handleSubmit = ( event ) => {
+		event.preventDefault();
+
+		if ( ! selectedOption || ! details.length || state.isSubmitted || state.isSubmitting ) {
+			return;
+		}
+
+		dispatch( { status: 'submitting' } );
+
+		apiFetch( {
+			path: addQueryArgs( '/wp/v2/wporg-pattern-flag' ),
+			method: 'POST',
+			data: {
+				parent: postId,
+				'wporg-pattern-flag-reason': selectedOption,
+				excerpt: details,
+			},
+		} )
+			.then( () => {
+				dispatch( { status: 'submitted' } );
+			} )
+			.catch( ( err ) => {
+				dispatch( {
+					status: 'error',
+					message: err.message,
+				} );
+			} );
+	};
+
+	const handleClose = () => {
+		onClose( state.isSubmitted );
+	};
+
+	const mappedReasons = useMemo( () => {
```

Comment:
Could we do this in the useSelect request?

Reply:
We could but I thought it makes more sense to keep the `reasons` exported from the state in their original form. This mapping is very specific to the radio control. If this were the innermost child component, I would have fewer reservations.

Reply:
Right, I didn't mean to change this in the state, but to do something like this inside the `useSelect` callback:

```js
const reasons = getPatternFlagReasons();

// map reasons for radiocontrol

return {
	isLoading: isLoadingPatternFlagReasons(),
	reasons: mappedReasons,
};
```

Reply:
Yea, I don't mean in the state exactly but I see `useSelect` as a representation of the state. For me, it would be akin to mapping for a component inside the traditional `function mapStateToProps(state, ownProps?)` function and you would never do that.

Reply:
I think it's fairly common to reshape the data from `useSelect` and `mapStateToProps` for whatever the component actually needs. In this case, it would remove the need for `useMemo`. You could call the return value `reasonList` or something to make it clear it's not a direct state value?